server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java

```diff
@@ -0,0 +1,160 @@
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Service class that handles segment replication requests from replica shards.
+ * Typically, the "source" is a primary shard. This code executes on the source node.
+ *
+ * @opensearch.internal
+ */
+public class SegmentReplicationSourceService {
+
+    private static final Logger logger = LogManager.getLogger(SegmentReplicationSourceService.class);
+
+    /**
+     * Internal actions used by the segment replication source service on the primary shard
+     *
+     * @opensearch.internal
+     */
+    public static class Actions {
+        public static final String GET_CHECKPOINT_INFO = "internal:index/shard/replication/get_checkpoint_info";
+        public static final String GET_SEGMENT_FILES = "internal:index/shard/replication/get_segment_files";
+    }
+
+    private final Map<ReplicationCheckpoint, CopyState> copyStateMap;
+    private final TransportService transportService;
+    private final IndicesService indicesService;
+
+    // TODO mark this as injected and bind in Node
+    public SegmentReplicationSourceService(TransportService transportService, IndicesService indicesService) {
+        copyStateMap = Collections.synchronizedMap(new HashMap<>());
+        this.transportService = transportService;
+        this.indicesService = indicesService;
+
+        transportService.registerRequestHandler(
+            Actions.GET_CHECKPOINT_INFO,
+            ThreadPool.Names.GENERIC,
+            CheckpointInfoRequest::new,
+            new CheckpointInfoRequestHandler()
+        );
+        transportService.registerRequestHandler(
+            Actions.GET_SEGMENT_FILES,
+            ThreadPool.Names.GENERIC,
+            GetSegmentFilesRequest::new,
+            new GetSegmentFilesRequestHandler()
+        );
+    }
+
+    private class CheckpointInfoRequestHandler implements TransportRequestHandler<CheckpointInfoRequest> {
+        @Override
+        public void messageReceived(CheckpointInfoRequest request, TransportChannel channel, Task task) throws Exception {
+            final ReplicationCheckpoint checkpoint = request.getCheckpoint();
+            logger.trace("Received request for checkpoint {}", checkpoint);
+            final CopyState copyState = getCachedCopyState(checkpoint);
+            channel.sendResponse(
+                new CheckpointInfoResponse(
+                    copyState.getCheckpoint(),
+                    copyState.getMetadataSnapshot(),
+                    copyState.getInfosBytes(),
+                    copyState.getPendingDeleteFiles()
+                )
+            );
+        }
+    }
+
+    class GetSegmentFilesRequestHandler implements TransportRequestHandler<GetSegmentFilesRequest> {
+        @Override
+        public void messageReceived(GetSegmentFilesRequest request, TransportChannel channel, Task task) throws Exception {
+            if (isInCopyStateMap(request.getCheckpoint())) {
+                // TODO send files
```

Comment:
I don't think we have a separate issue to actually implement the copy here.  Should we leave this issue open until that happens? I'm ok with this shipping incrementally.

Reply:
Ah fair, I'll leave #3322 open